include/Core/UI/Setting/EmailLogSetting.php

```diff
@@ -0,0 +1,113 @@
+<?php
+namespace EmailLog\Core\UI\Setting;
+use EmailLog\Core\UI\Setting\SettingField;
+
+class EmailLogSetting extends Setting {
+
+	/**
+	 * Implement `initialize()` method.
+	 */
+	protected function initialize() {
+		$this->section->id          = 'email-log';
+		$this->section->title       = __( 'Plugin Settings', 'email-log' );
+		$this->section->option_name = 'el_email_log';
+
+		$this->load();
+	}
+
+	/**
+	 * Implement `get_fields()` method.
+	 *
+	 * @return array
+	 */
+	public function get_fields() {
+		$fields = array();
+
+		$email_log_fields = array(
+			'allowed_user_roles'  => __( 'Allowed User Roles', 'email-log' ),
+		);
+
+		foreach ( $email_log_fields as $field_id => $label ) {
+			$field           = new SettingField();
+			$field->id       = $field_id;
+			$field->title    = $label;
+			$field->args     = array( 'id' => $field_id );
+			$field->callback = array( $this, 'render_email_log_capability_field' );
+
+			$fields[] = $field;
+		}
+
+		return $fields;
+	}
+
+	/**
+	 * Implement `render()` method.
+	 */
+	public function render() {
+	?>
+		<p><?php _e( 'Users with the following <strong>User Roles</strong> can view Email Logs. The default User Role is \'<strong>administrator</strong>\'.', 'email-log' ); ?></p>
+	<?php
+	}
+
+	/**
+	 * Implement `sanitize()` method.
+	 *
+	 * @param mixed $values {@inheritDoc}
+	 *
+	 * @return mixed $values {@inheritDoc}
+	 */
+	public function sanitize( $values ) {
+		if ( ! is_array( $values ) ) {
+			return array();
+		}
+
+		foreach ( $values as $key => $value ) {
+			if ( $key === 'allowed_user_roles' ) {
+				$values[ $key ] = array_map( 'sanitize_text_field', $values[ $key ] );
+			}
+		}
+
+		return $values;
+	}
+
+	/**
+	 * Renders the Capability field to set Capability.
+	 *
+	 * @param array $args
+	 */
+	public function render_email_log_capability_field( $args ) {
+		$option          = $this->get_value();
+		$option          = $option[ $args['id'] ];
+		$available_roles = get_editable_roles();
+		foreach( $available_roles as $role ) {
+			if ( trim( $role['name'] ) === 'Administrator' ) {
+	?>
+			<p><input type="checkbox" name="<?php echo esc_attr( $this->section->option_name . '[' . $args['id'] . '][]' ); ?>" value="<?php echo trim( $role['name'] ); ?>" checked="checked" /> <?php echo trim( $role['name'] ); ?>
```

Comment:
Let's make the `Administrator` field as read-only and always checked.

Also you can hardcode the admin checkbox and then remove the admin role from `$available_roles` array and then just iterate it without having an extra conditional block.

Reply:
> Let's make the Administrator field as read-only and always checked.

I did it initially. But then I realized that read-only fields are not submitted in form POST. So I figured a [work-around](https://stackoverflow.com/a/6241951/5569979). But, that looked like I'm complicating things. So I preferred to stick to the current implementation. Kindly take a look and let me know, if you prefer to have hidden input fields to achieve read-only.

Reply:
You can just make the filed as disabled and then add the admin role in the backend before saving it.

This will not require the workaround.

Also this will make sure that even if someone removes admin and submits the form (by tampering with the request) it will still work.

Reply:
> You can just make the filed as disabled and then add the admin role in the backend before saving it. This will not require the workaround.

Okay, will change it.

> Also this will make sure that even if someone removes admin and submits the form (by tampering with the request) it will still work.

The Administrator checkbox will still be checked when saving, even if the User removes it in the front-end. (That's why I have included a foot note in the Setting ;) )

Reply:
> You can just make the filed as disabled and then add the admin role in the backend before saving it.

I see a caveat in using this method. The `sanitize()` method will return empty array when the $values are empty. Assuming we use disable the Administrator checkbox field, and when no other option is selected by the user, the $values will be empty. Since we use the same `sanitize()` method for both the Settings field, it is hard to determine when to send empty array and when to send array( 'Administrator' ) in order to save the Settings values in DB.

````
if ( ! is_array( $values ) ) {
	return array();
}
````

So, we have two choices.

- We continue to use existing method.
- We split the `sanitize()` method now to avoid the above caveat.

Kindly advise.

Reply:
Let's split the `sanitize` method then.

Reply:
> Let's split the `sanitize` method then.

I think we will have to modify this class first, in order to achieve splitting the `sanitize()` method - `email-log/include/Core/UI/Setting/Setting.php`